src/sc_client/session.py

```diff
@@ -66,91 +84,110 @@ def _emit_callback(event_id: int, elems: list[int]) -> None:
-            url,
-            on_message=_on_message,
-            on_error=_on_error,
-            on_close=_on_close,
-        )
-        _ScClientSession.ws_app.run_forever()
+        if not _ScClientSession.ws_app or _ScClientSession.ws_app.url != url:
+            _ScClientSession.ws_app = websocket.WebSocketApp(
+                url,
+                on_message=_on_message,
+                on_error=_on_error,
+                on_close=_on_close,
+            )
+            logger.info(f"Sc-server socket: {_ScClientSession.ws_app.url}")
+        if not _ScClientSession.is_open:
+            try:
+                _on_open(_ScClientSession.ws_app)
+                logger.info("Set new connection")
+                _ScClientSession.ws_app.run_forever()
+            except websocket.WebSocketException as e:
+                _on_error(_ScClientSession.ws_app, e)
 
     client_thread = threading.Thread(target=run_in_thread, name="sc-client-session-thread")
     client_thread.start()
     time.sleep(SERVER_ESTABLISH_CONNECTION_TIME)
-    logger.debug("Trying to establish connection: %s", "successful" if is_connection_established() else "unsuccessful")
-
-
-def set_reconnection(url: str) -> None:
-    def run_in_thread():
-        while _ScClientSession.do_reconnect:
-            time.sleep(SERVER_ESTABLISH_CONNECTION_TIME)
-            if not is_connection_established():
-                logger.debug("No connection, trying to reconnect")
-                establish_connection(url)
-                if not is_connection_established():
-                    logger.debug("Still no connection, waiting")
-                    time.sleep(_ScClientSession.reconnecion_time)
-
-    reconnection_thread = threading.Thread(target=run_in_thread, name="reconnection-thread")
-    reconnection_thread.start()
-
-
-def disable_reconnection() -> None:
-    _ScClientSession.do_reconnect = False
 
 
 def close_connection() -> None:
     try:
         _ScClientSession.ws_app.close()
-    except AttributeError:
-        pass
-    _ScClientSession.ws_app = None
+    except AttributeError as e:
+        _on_error(_ScClientSession.ws_app, e)
     logger.debug("Disconnected")
 
 
 def receive_message(command_id: int) -> Response:
     response = None
-    while not response:
+    while not response and _ScClientSession.is_open:
         response = _ScClientSession.responses_dict.get(command_id)
         time.sleep(SERVER_ANSWER_CHECK_TIME)
     return response
 
 
-def _send_message(data: str) -> None:
-    _ScClientSession.ws_app.send(data)
-    logger.debug(f"Send: {data[:LOGGING_MAX_SIZE]}")
+def _send_message(data: str, retries: int, retry_delay: float, retry: int = 0) -> None:
+    try:
+        logger.debug(f"Send: {data[:LOGGING_MAX_SIZE]}")
+        _ScClientSession.ws_app.send(data)
+    except websocket.WebSocketConnectionClosedException as e:
+        _on_close(_ScClientSession.ws_app, 1, "Connection broken")
```

Comment:
I don't get it. We've already bound those functions to WebSocket client when you've created a WebSocket app, so those should be called automatically on any exception. Why are we calling it manually?

Reply:
No, that's not how it works. All exceptions are thrown only from the library code when something went wrong with request processing. Errors that were raised in client code should be thrown like this